gnucash/gnome/gnc-budget-view.c

```diff
@@ -1689,3 +1692,30 @@ gnc_budget_view_refresh (GncBudgetView *budget_view)
     LEAVE(" ");
 }
 
+
+/* this function returns TRUE if we must allow budget cell editing:
+   the budgets have been fixed already, or the user is ready to fix
+   budgets manually. */
+gboolean gnc_budget_allow_editing (GtkWidget *window)
+{
+    QofBook *book = gnc_get_current_book();
+
+    /* feature already set. allow budget editing to continue. */
+    if (gnc_features_check_used (book, GNC_FEATURE_BUDGET_UNREVERSED))
+        return TRUE;
+
+    /* feature not set. ask user before setting feature. */
+    if (!gnc_verify_dialog (GTK_WINDOW (window), FALSE, (_ ("You are editing \
+a budget cell manually. To proceed, we need to set a feature to prevent \
+budget signs errors. This is important to fix inconsistencies in budget \
+reports. This change will cause the data to be unreadable in 3.7 or earlier. \
+If you proceed, It is important to recheck all budgets to ensure the amounts \
+signs are correct.\n\nDo you wish to proceed to set the feature, and allow \
+editing?"))))
+        /* user declined feature. abort editing. */
+        return FALSE;
+
+    /* accept feature. allow editing. */
```

Comment:
That's the confusing part: https://github.com/Gnucash/gnucash/pull/585#pullrequestreview-303462160 - then the user is expected to fix signs manually.

Reply:
FWIW the approach implemented in this branch is as suggested by @gjanssens is to allow the intermediate state - budgets are read-only until auto-fix is allowed, or budget is unlocked for editing. This means both branches of code need to be maintained forever.

My own preference was to remove the intermediate state and enforce auto-fix; this would enable old code handling reversed budgets to be removed.

I think this branch works well.

Reply:
https://code.gnucash.org/logs/2021/01/24.html#T19:17:42

Do we:
1. consider merging this branch, i.e. offer scrubbing; if declined then budgets are read-only until manually fixed?
2. enforce scrubbing and advise manual review of all budgets?
3. offer scrubbing + readonly budgets for 4.x; and enforce scrubbing and remove read-only budgets for 5.x

Reply:
I'm for #2.

But I think there's a documentation component to this as well: Will anything described in the T&CG [budget chapter](https://www.gnucash.org/docs/v4/C/gnucash-guide/chapter_budgets.html) change? Does it need to explain the impact of different values of the sign reversal option to keep users from being confused?

Reply:
Option 2 is my preference too -- but Geert prefers 1.

I'll review and add another commit to illustrate simplicity of option 2.

The budget chapter could be amended slightly; it'd mean the budget amount signs _will_ be similar to the account signs: e.g. if they expect income accounts to be negative, then budget income amounts will be negative too.

Reply:
I think my opinion weighs less today given my limited availability...

Having said that, I  do prefer some form of safety net for unsuspecting users. It doesn't have to be as suggested in option 1 though.

Here's the use case I want to cover:
- User installs gnucash 5.x without having read the release notes about sign fixes in budgets. User as a sql based book, and was careless enough to not make a backup.
- User opens  book for the first time and is confronted with budget scrubbing.
- Imagine it goes horribly wrong for whatever reason (a corner case we didn't consider for example). Sql backend means changes are applied immediately so there's no easy rollback.
- Lots of user frustration and support time spent to fix this up again.

Of course the user *should* have taken that backup and have read the release notes.

We can make user's lives somewhat more pleasant by informing the user at startup a budget fix is about to happen if user continues with loading this book (even if that will be enforced), we could offer the option to not open the book instead (and to make that backup before trying again). That will give the user a choice:
- continue and take the "risk"
- bail out, make that forgotten backup first and then try again
- bail out and stay on 4.x for some time longer to hear the experience of others first
The dialog itself doesn't have to offer 3 options though. It can shortly explain what's about to happen, make mention of the importance of a good backup and finally ask if the user wants to proceed or quit.

IMHO that could  become a more generic strategy on major version upgrades in the future for any data format conversion we do.

Reply:
I like that. Gramps does something like "This new version of Gramps uses a new database format and we need to upgrade it. If you don't have a good backup, please make one now before continuing." with Cancel and OK buttons. Cancel doesn't do the upgrade and closes the database. 

The difference for GnuCash in doing something like that is that we tend to do scrubs when using some feature and it might be well into a session. We might need a backend-dependent action in that case where if it's XML we can just do an autosave but if it's SQL the user has to stop and either copy the SQLite3 file or do whatever backup procedure is correct for the server in question. Cancel in our case might just cancel whatever action triggered the scrub rather than closing the file because the rest of the program might work OK.

Where it gets interesting is when the user accepts the first scrub and then encounters a second one and doesn't want to accept it. Depending on what are the first versions that support each feature there's a potential that they'll end up in a situation where they can't go back to the pre-scrub of the second feature because it's blocked by the minimum version of the first.

Reply:
[groan] More accidental complexity.

So: we'd need to overhaul the `gnc_features` feature, ensure all scrubs are centralised and triggered upon load (cf upon first use) and wrap around them to offer the bail-out mechanism?

Reply:
Yes, I think our current strategy to only do conversions the moment a new feature is used for the first time makes our compatibility story very complicated. I have considered it more than once in the past to rework this into the scenario you suggest: run any needed conversion at file load time. We can still keep the "features" concept and only set features that are actually used. For example if a user doesn't have budgets there's no need to set the sign reversal fix flag so older versions of gnucash will continue to work. The only difference is the timing of the conversions.

Reply:
I had a thought how to add feature only if required: add `set_budget_feature (book)` - will count budgets; if >=1 then will set feature, otherwise remove feature:
1. any budget modifier (i.e. create/delete) will call it
2. the budget scrubber, when trigged on file load, will also call it.